### PR TITLE
[AutoDiff] [SR-14218] Correctly propagate tangent vectors of inout parameters from functions with multiple basic blocks

### PR DESCRIPTION
**Pull request by BradLarson**
Pullback generation for functions featuring active inout parameters and multiple basic blocks currently does not correctly propagate tangent vectors for the inout parameters. Changes to the tangent vectors of the inout parameters are not being copied back at the end of pullbacks involving control flow. As a workaround, to get correct tangent vectors you currently need to manually copy tangent vectors to an outside holding variable, then manually copy those tangent vectors back in before the function call site.

This pull request adds the missing copies of tangent vectors for inout parameters at the end of pullbacks, leading to correct tangent vectors around functions using inout parameters and control flow. Tests have been added for common permutations of these functions that previously produced incorrect tangent vectors.

In one large differentiable physics simulator with a complex model, this eliminates 80 lines of code for manual tangent vector copies. It also improves gradient descent throughput by 80% due to the removal of overhead from these explicit tangent vector copies.

Most of the credit for this fix belongs to @dan-zheng, who provided the original code to correctly copy these tangent vectors and worked with me to refine the fix.

Resolves [SR-14053](https://bugs.swift.org/browse/SR-14053) and [SR-14218](https://bugs.swift.org/browse/SR-14218).

### CODE REVIEW

**Comment by BradLarson** [comment]
@swift-ci Please test

**Comment by BradLarson** [comment]
@swift-ci Please test macOS platform

**Review by dan-zheng — APPROVED** [review]
Looks good!

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp [hunk 1]
```diff
@@ -2052,6 +2052,8 @@ bool PullbackCloner::Implementation::run() {
   SmallVector<SILValue, 8> retElts;
   // This vector will contain all indirect parameter adjoint buffers.
   SmallVector<SILValue, 4> indParamAdjoints;
+  // This vector will identify the locations where initialization is needed.
+  SmallVector<bool, 8> outputsToInitialize;
```
Comment (rxwei) [inline lib/SILOptimizer/Differentiation/PullbackCloner.cpp:2056]: Is it possible to use a `SmallBitVector` instead?

**Review by rxwei — APPROVED** [review]

File: lib/SILOptimizer/Differentiation/PullbackCloner.cpp (hunk 1 above)
Comment (BradLarson) [reply lib/SILOptimizer/Differentiation/PullbackCloner.cpp]: Thanks for the suggestion, I've gone ahead and done so.

**Comment by BradLarson** [comment]
@swift-ci Please test

**Comment by BradLarson** [comment]
@swift-ci Please test Windows platform

**Comment by BradLarson** [comment]
@swift-ci Please test Windows platform

**Comment by BradLarson** [comment]
@swift-ci Please test

**Comment by BradLarson** [comment]
@swift-ci Please test macOS platform

**Comment by BradLarson** [comment]
@swift-ci Please test Windows platform

**Comment by BradLarson** [comment]
@swift-ci Please test macOS

**Comment by BradLarson** [comment]
@swift-ci Please test macOS

**Comment by BradLarson** [comment]
@swift-ci Please test macOS platform

**Comment by BradLarson** [comment]
@swift-ci Please test macOS platform

**Comment by BradLarson** [comment]
@swift-ci Please test macOS platform

**Comment by BradLarson** [comment]
@swift-ci Please test macOS platform

**Pull request merged**